Source/Core/Core/AchievementManager.cpp

```diff
@@ -1016,694 +520,461 @@ void AchievementManager::FilereaderClose(void* file_handle)
-                        OSD::Duration::VERY_LONG, OSD::Color::YELLOW);
-      }
-      else
-      {
-        OSD::AddMessage(fmt::format("Scored {} on leaderboard: {}", value.data(),
-                                    m_game_data.leaderboards[ix].title),
-                        OSD::Duration::VERY_LONG, OSD::Color::YELLOW);
-      }
-      m_queue.EmplaceItem([this, event_id] {
-        FetchBoardInfo(event_id);
-        m_update_callback();
-      });
-      break;
-    }
+    auto value = system->GetMMU().HostTryReadU8(threadguard, address + num_read,
+                                                PowerPC::RequestedAddressSpace::Physical);
+    if (!value.has_value())
+      return num_read;
+    buffer[num_read] = value.value().value;
   }
-  ERROR_LOG_FMT(ACHIEVEMENTS, "Invalid leaderboard triggered event with id {}.", event_id);
+  return num_bytes;
 }
 
-// Every RetroAchievements API call, with only a partial exception for fetch_image, follows
-// the same design pattern (here, X is the name of the call):
-//   Create a specific rc_api_X_request_t struct and populate with the necessary values
-//   Call rc_api_init_X_request to convert this into a generic rc_api_request_t struct
-//   Perform the HTTP request using the url and post_data in the rc_api_request_t struct
-//   Call rc_api_process_X_response to convert the raw string HTTP response into a
-//     rc_api_X_response_t struct
-//   Use the data in the rc_api_X_response_t struct as needed
-//   Call rc_api_destroy_X_response when finished with the response struct to free memory
-template <typename RcRequest, typename RcResponse>
-AchievementManager::ResponseType AchievementManager::Request(
-    RcRequest rc_request, RcResponse* rc_response,
-    const std::function<int(rc_api_request_t*, const RcRequest*)>& init_request,
-    const std::function<int(RcResponse*, const char*)>& process_response)
+void AchievementManager::FetchBadge(AchievementManager::BadgeStatus* badge, u32 badge_type,
+                                    const AchievementManager::BadgeNameFunction& function,
+                                    const UpdatedItems& callback_data)
 {
-  rc_api_request_t api_request;
-  Common::HttpRequest http_request;
-  if (init_request(&api_request, &rc_request) != RC_OK || !api_request.post_data)
+  if (!m_client || !IsLoggedIn() || !Config::Get(Config::RA_BADGES_ENABLED))
   {
-    ERROR_LOG_FMT(ACHIEVEMENTS, "Invalid API request.");
-    return ResponseType::INVALID_REQUEST;
+    m_update_callback(callback_data);
+    return;
   }
-  auto http_response = http_request.Post(api_request.url, api_request.post_data);
-  rc_api_destroy_request(&api_request);
-  if (http_response.has_value() && http_response->size() > 0)
-  {
-    const std::string response_str(http_response->begin(), http_response->end());
-    if (process_response(rc_response, response_str.c_str()) != RC_OK)
+
+  m_image_queue.EmplaceItem([this, badge, badge_type, function, callback_data] {
```

Comment:
`function` should be moved instead of copied here  (though this function `FetchBadge` will have to pass by value instead of reference).

Similar thing for `callback_data` which is a fairly large blob of data.